File: ios/StoryTeller/Model/GraphNode.swift
import Foundation

struct GraphNode: Codable, Identifiable {
    let nodeId: String
    let text: String
    let choices: [Choice]
    let sceneId: String
    let locationId: String
    let participantIds: [String]
    let authoritativeRefs: [String]
    let ending: String?
    var id: String { nodeId }
    var isEnding: Bool { ending != nil }
    var displayLines: [String] { text.split(separator: "\n").map(String.init).filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty } }
    enum CodingKeys: String, CodingKey {
        case nodeId = "node_id", text, choices, sceneId = "scene_id", locationId = "location_id"
        case participantIds = "participant_ids", authoritativeRefs = "authoritative_refs", ending
    }
    init(nodeId: String, text: String, choices: [Choice], sceneId: String = "",
         locationId: String = "", participantIds: [String] = [], authoritativeRefs: [String] = [],
         ending: String? = nil) {
        self.nodeId=nodeId; self.text=text; self.choices=choices; self.sceneId=sceneId
        self.locationId=locationId; self.participantIds=participantIds
        self.authoritativeRefs=authoritativeRefs; self.ending=ending
    }
}

struct Choice: Codable, Identifiable {
    let choiceId: String
    let text: String
    let targetNode: String
    let routeId: String
    let setsFlags, requiresFlags: [String]
    var id: String { choiceId }
    var choiceText: String { text }
    enum CodingKeys: String, CodingKey {
        case choiceId = "choice_id", text, targetNode = "target_node", routeId = "route_id"
        case setsFlags = "sets_flags", requiresFlags = "requires_flags"
    }
    init(choiceId: String, choiceText: String, targetNode: String, setsFlags: [String] = [],
         requiresFlags: [String] = [], routeId: String = "") {
        self.choiceId=choiceId; self.text=choiceText; self.targetNode=targetNode
        self.setsFlags=setsFlags; self.requiresFlags=requiresFlags; self.routeId=routeId
    }
    func isAvailable(activeFlags: Set<String>) -> Bool { requiresFlags.allSatisfy(activeFlags.contains) }
    func isAvailable(activeFlags: [String: Bool]) -> Bool { requiresFlags.allSatisfy { activeFlags[$0] == true } }
}

struct StorySession {
    let nodes: [String: GraphNode]
    var state: SaveState
    var current: GraphNode { nodes[state.currentNode] ?? { fatalError("Unknown node") }() }
    func availableChoices() -> [Choice] { current.choices.filter { $0.isAvailable(activeFlags: Set(state.flags.filter(\.value).map(\.key))) } }
    mutating func choose(_ id: String) throws {
        guard let choice = availableChoices().first(where: { $0.choiceId == id }) else { throw SessionError.unavailable }
        guard nodes[choice.targetNode] != nil else { throw SessionError.missingTarget }
        choice.setsFlags.forEach { state.flags[$0] = true }; state.currentNode = choice.targetNode
        if !state.visitedNodes.contains(choice.targetNode) { state.visitedNodes.append(choice.targetNode) }
    }
    enum SessionError: Error { case unavailable, missingTarget }
}

File: droid/app/src/main/java/com/storyteller/droid/model/GraphNode.kt
package com.storyteller.droid.model

/** Frozen v2 graph model. */
data class GraphNode(
    val nodeId: String,
    val text: String,
    val choices: List<Choice>,
    val chapter: Int = 1,
    val sceneType: String = "narrative",
    val mood: String = "",
    val sceneId: String = "",
    val locationId: String = "",
    val participantIds: List<String> = emptyList(),
    val authoritativeRefs: List<String> = emptyList(),
    val ending: String? = null,
    val isEnding: Boolean = ending != null,
) {
    val displayLines get() = text.split("\n").filter { it.isNotBlank() }
}

data class Choice(
    val choiceId: String,
    val choiceText: String,
    val targetNode: String,
    val setsFlags: List<String> = emptyList(),
    val requiresFlags: List<String> = emptyList(),
    val routeId: String = "",
) {
    fun isAvailable(activeFlags: Set<String>) = requiresFlags.all(activeFlags::contains)
    fun isAvailable(activeFlags: Map<String, Boolean>) =
        requiresFlags.all { activeFlags[it] == true }
}

/** Pure deterministic state transition shared by UI and tests. */
class StorySession(private val nodes: Map<String, GraphNode>, entryNode: String,
                   val state: SaveState) {
    init { require(entryNode in nodes); if (state.currentNode.isEmpty()) state.currentNode = entryNode }
    val current: GraphNode get() = nodes[state.currentNode] ?: error("Unknown current node")
    fun availableChoices() = current.choices.filter { it.isAvailable(state.flags.filterValues { v -> v }.keys) }
    fun choose(choiceId: String): GraphNode {
        val choice = availableChoices().singleOrNull { it.choiceId == choiceId }
            ?: throw IllegalArgumentException("CHOICE_UNAVAILABLE")
        require(choice.targetNode in nodes) { "CHOICE_TARGET_MISSING" }
        choice.setsFlags.forEach { state.flags[it] = true }
        state.currentNode = choice.targetNode
        if (choice.targetNode !in state.visitedNodes) state.visitedNodes.add(choice.targetNode)
        return current
    }
}
